cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+	"github.com/azure/azure-dev/cli/azd/pkg/output"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/bash"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools/powershell"
+	"golang.org/x/exp/slices"
+)
+
+// Generic action function that may return an error
+type InvokeFn func() error
+
+// The type of command hooks. Supported values are 'pre' and 'post'
+type HookType string
+
+const (
+	// Executes pre-command hooks
+	HookTypePre HookType = "pre"
+	// Execute post-command hooks
+	HookTypePost HookType = "post"
+)
+
+// CommandHooks enable support to invoke integration scripts before & after commands
+// Scripts can be invoked at the project or service level or
+type CommandHooks struct {
+	commandRunner exec.CommandRunner
+	console       input.Console
+	cwd           string
+	scripts       map[string]*ScriptConfig
+	envVars       []string
+}
+
+// NewCommandHooks creates a new instance of CommandHooks
+func NewCommandHooks(
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	scripts map[string]*ScriptConfig,
+	cwd string,
+	envVars []string,
+) *CommandHooks {
+	if cwd == "" {
+		osWd, err := os.Getwd()
+		if err != nil {
+			panic(err)
+		}
+
+		cwd = osWd
+	}
+
+	return &CommandHooks{
+		commandRunner: commandRunner,
+		console:       console,
+		cwd:           cwd,
+		scripts:       scripts,
+		envVars:       envVars,
+	}
+}
+
+// Invokes an action run runs any registered pre or post script hooks for the specified command.
+func (h *CommandHooks) Invoke(ctx context.Context, commands []string, actionFn InvokeFn) error {
+	err := h.RunScripts(ctx, HookTypePre, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	err = actionFn()
+	if err != nil {
+		return err
+	}
+
+	err = h.RunScripts(ctx, HookTypePost, commands)
+	if err != nil {
+		return fmt.Errorf("failing running pre command hooks: %w", err)
+	}
+
+	return nil
+}
+
+// Invokes any registered script hooks for the specified hook type and command.
+func (h *CommandHooks) RunScripts(ctx context.Context, hookType HookType, commands []string) error {
+	scripts := h.getScriptsForHook(hookType, commands)
+	for _, scriptConfig := range scripts {
+		err := h.execScript(ctx, scriptConfig)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Gets the script to execute based on the script configuration values
+// For inline scripts this will also create a temporary script file to execute
+func (h *CommandHooks) GetScript(scriptConfig *ScriptConfig) (tools.Script, error) {
+	if scriptConfig.Location == "" {
+		if scriptConfig.Path != "" {
+			scriptConfig.Location = ScriptLocationPath
+		} else if scriptConfig.Script != "" {
+			scriptConfig.Location = ScriptLocationInline
+		}
+	}
+
+	if scriptConfig.Location == ScriptLocationInline {
+		tempScript, err := createTempScript(scriptConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		scriptConfig.Path = tempScript
```

Comment:
Does this work as expected? `createTempScript` returns `file.Name()` whereas the `Path` here is expected to be relative to `cwd`.

Reply:
`file.Name()` returns the relative path from the `cwd` in this case.